torchvision/models/quantization/resnet.py

```diff
@@ -0,0 +1,239 @@
+
+        out = self.add_relu.add_relu(out, identity)
+
+        return out
+
+    def fuse_model(self):
+        torch.quantization.fuse_modules(self, [['conv1', 'bn1', 'relu'],
+                                            ['conv2', 'bn2']], inplace=True)
+        if self.downsample:
+            torch.quantization.fuse_modules(self.downsample, ['0', '1'], inplace=True)
+
+
+class QuantizableBottleneck(Bottleneck):
+    def __init__(self, *args, **kwargs):
+        super(QuantizableBottleneck, self).__init__(*args, **kwargs)
+        self.skip_add_relu = nn.quantized.FloatFunctional()
+        self.relu1 = nn.ReLU(inplace=False)
+        self.relu2 = nn.ReLU(inplace=False)
+
+    def forward(self, x):
+        identity = x
+        out = self.conv1(x)
+        out = self.bn1(out)
+        out = self.relu1(out)
+        out = self.conv2(out)
+        out = self.bn2(out)
+        out = self.relu2(out)
+
+        out = self.conv3(out)
+        out = self.bn3(out)
+
+        if self.downsample is not None:
+            identity = self.downsample(x)
+        out = self.skip_add_relu.add_relu(out, identity)
+
+        return out
+
+    def fuse_model(self):
+        fuse_modules(self, [['conv1', 'bn1', 'relu1'],
+                        ['conv2', 'bn2', 'relu2'],
+                        ['conv3', 'bn3']], inplace=True)
+        if self.downsample:
+            torch.quantization.fuse_modules(self.downsample, ['0', '1'], inplace=True)
+
+
+class QuantizableResNet(ResNet):
+
+    def __init__(self, *args, **kwargs):
+        super(QuantizableResNet, self).__init__(*args, **kwargs)
+
+        self.quant = torch.quantization.QuantStub()
+        self.dequant = torch.quantization.DeQuantStub()
+
+    def forward(self, x):
+        x = self.quant(x)
+        # Ensure scriptability
+        # super(QuantizableResNet,self).forward(x)
+        # is not scriptable
+        x = self._forward(x)
+        x = self.dequant(x)
+        return x
+
+    def fuse_model(self):
+        r"""Fuse conv/bn/relu modules in resnet models
+
+        Fuse conv+bn+relu/ Conv+relu/conv+Bn modules to prepare for quantization.
+        Model is modified in place.  Note that this operation does not change numerics
+        and the model after modification is in floating point
+        """
+
+        fuse_modules(self, ['conv1', 'bn1', 'relu'], inplace=True)
+        for m in self.modules():
+            if type(m) == QuantizableBottleneck or type(m) == QuantizableBasicBlock:
+                m.fuse_model()
+
+
+def _resnet(arch, block, layers, pretrained_float_model, progress, **kwargs):
+    model = QuantizableResNet(block, layers, **kwargs)
+    _replace_relu(model)
+    if pretrained_float_model:
+        state_dict = load_state_dict_from_url(model_urls[arch],
+                                              progress=progress)
+        model.load_state_dict(state_dict)
+    return model
+
+
+def resnet18(pretrained_float_model=False, progress=True, **kwargs):
```

Comment:
Are we going to be providing pre-trained model weights for all those models?

If not, I'd recommend only keeping the ones we are sure to get the weights for. We can always add more models afterwards, but we might not have the bandwidth to train all the models we have in this list in the near future.

I'd recommend only keeping only the models that are going to be tested.

Thoughts?

Reply:
Sorry, had missed your previous comment: This PR allows for quantizing any floating point model specified in resnet/mobilenet.py.

Reply:
see my previous comment. The fact that the model has a pre-trained weight associated with it plays an important role for it being present in torchvision or not, and we should rather start with only a few models than to let many models without pre-trained weights for a long time

Reply:
All the models supported have pretrained weights, it is just that we start with floating point weights and provide code to generate quantized models. I can explain in more detail in a chat.